### PR TITLE
Flatten Density Speedup + Time Report Bug Fix

### LINKED ISSUE #138: Can FFTW array allocation be done only once per inset? Can the FFTW plan be recycled?
Unfortunately, `flatten_density()` is quite slow. I am wondering whether some speed could be gained by allocating memory for `grid_flux*_init` before the first iteration of `flatten_density()` and then be kept alive until the integration is finished for this inset. At the same points in the code, the plans could perhaps be created and destroyed only once per inset.

### PR DESCRIPTION
**Pull request by nihalzp**
Closes #138

### CODE REVIEW

**Comment by adisidev** [comment]
@nihalzp do you think you could change the last commit such that it is compatible with a dynamic grid?

File: src/inset_state/flatten_density.cpp [hunk 1]
```diff
@@ -453,24 +447,26 @@ void InsetState::flatten_density_with_node_vertices()
...
-        t + 0.5 * delta_t,
-        grid_fluxx_init,
-        grid_fluxy_init,
-        rho_ft_,
-        rho_init_,
-        &grid_vx,
-        &grid_vy,
-        lx_,
-        ly_);
+      std::function<double(unsigned int, unsigned int, char)> get_velocity =
```
Comment (mgastner) [inline src/inset_state/flatten_density.cpp:450]: The semantic difference between `cal_velocity` and `get_velocity` is not immediately obvious. Would `cal_velocity_at_half_time` be better than `get_velocity()`?

**Review by mgastner — CHANGES_REQUESTED** [review]
My main request for changes is the same as Adi's: resizing the `FTReal2d` arrays holds more promise for shorter runtimes than fewer memory allocations of a fixed array size. Adi is already working on that project. In this respect, I would return to the original and make `grid_fluxx` and `grid_fluxy` local variables again. The rest looks good at first glance (apart from a minor suggestion about the name of a lambda function, see inline comment).

**Comment by nihalzp** [comment]
Thank you for the review Prof. Gastner and Adi.

I think the current version is compatible with Adi's change. In Adi's pull request, in `inset_state.cpp ` line 411, we can add the following lines to make the flux grid dynamic:

```
ref_to_fluxx_init()->allocate(lx_, ly_);
ref_to_fluxy_init()->allocate(lx_, ly_);
make_fftw_plans_for_flux();
```

File: src/inset_state/flatten_density.cpp (hunk 1 above)
Comment (nihalzp) [reply src/inset_state/flatten_density.cpp]: Turns out I missed renaming `get_velocity` to `cal_velocity`. Both serve the same function. So, I feel it is best to let them have the same name.

**Comment by nihalzp** [comment]
The content of this PR has been merged to the 'merge-pr' branch.

**Pull request closed without merging**